File: dart/interview_preparation/bin/learning/singly_linked_list.dart
class SinglyLinkedListNode {
  SinglyLinkedListNode({
    this.data = -1,
    this.next,
  });

  final int data;
  SinglyLinkedListNode? next;

  @override
  String toString() => "$data->$next";
}

void singlyLinkedList() {
  print("* Singly Linked List *");
  final singlyLinkedList = SingleLinkedList();
  singlyLinkedList.insert(0, 29);
  singlyLinkedList.unShift(1);
  singlyLinkedList.remove(2);
  singlyLinkedList.push(3);
  singlyLinkedList.push(4);
  singlyLinkedList.pop();
  singlyLinkedList.push(5);
  singlyLinkedList.remove(6);
  singlyLinkedList.push(7);
  singlyLinkedList.pop();
  singlyLinkedList.push(8);
  singlyLinkedList.remove(9);
  singlyLinkedList.push(10);
  singlyLinkedList.remove(11);
  singlyLinkedList.remove(12);
  singlyLinkedList.push(13);
  singlyLinkedList.push(14);
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.push(15);
  singlyLinkedList.push(16);
  singlyLinkedList.pop();
  singlyLinkedList.unShift(17);
  singlyLinkedList.unShift(18);
  singlyLinkedList.pop();
  singlyLinkedList.unShift(19);
  singlyLinkedList.unShift(20);
  singlyLinkedList.remove(21);
  singlyLinkedList.remove(22);
  singlyLinkedList.get(100);
  singlyLinkedList.get(1);
  singlyLinkedList.get(0);
  singlyLinkedList.get(8);
  singlyLinkedList.set(8, 23);
  singlyLinkedList.set(9, 24);
  singlyLinkedList.push(25);
  singlyLinkedList.push(26);
  singlyLinkedList.push(27);
  singlyLinkedList.set(9, 28);
  singlyLinkedList.insert(5, 30);
  singlyLinkedList.insert(0, 31);
  singlyLinkedList.insert(1, 32);
  singlyLinkedList.insert(16, 33);
  singlyLinkedList.insert(16, 34);
  singlyLinkedList.insert(18, 35);
  singlyLinkedList.insert(20, 35);
  singlyLinkedList.removeAt(20);
  singlyLinkedList.insert(19, 36);
  singlyLinkedList.removeAt(0);
  singlyLinkedList.removeAt(19);
  singlyLinkedList.reverse();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.pop();
  singlyLinkedList.shift();
  singlyLinkedList.reverse();
  singlyLinkedList.pop();
  singlyLinkedList.pop();
  singlyLinkedList.reverse();
  singlyLinkedList.pop();
  singlyLinkedList.reverse();
  print("* End Singly Linked List *");
}

class SingleLinkedList {
  SinglyLinkedListNode? head;

  void push(int data) {
    final newNode = SinglyLinkedListNode(
      data: data,
    );
    if (head == null) {
      head = newNode;
    } else {
      var temp = head;
      while (temp?.next != null) {
        temp = temp?.next;
      }
      temp?.next = newNode;
    }
    print("Pushed $newNode in $head");
  }

  void pop() {
    if (head == null) {
      print("Head is null.");
      return;
    }
    var temp = head;
    SinglyLinkedListNode? prev;
    while (temp?.next != null) {
      prev = temp;
      temp = temp?.next;
    }
    if (prev == null && temp?.next == null) {
      head = null;
    } else {
      prev?.next = null;
    }
    print("Popped last element $head");
  }

  void shift() {
    if (head == null) {
      print("Head is null.");
      return;
    }
    head = head?.next;
    print("Shifted head $head");
  }

  void unShift(int data) {
    final newNode = SinglyLinkedListNode(
      data: data,
    );
    if (head == null) {
      head = newNode;
    } else {
      newNode.next = head;
      head = newNode;
    }
    print("Inserted ${SinglyLinkedListNode(
      data: data,
    )} at the beginning in $head");
  }

  void remove(int data) {
    if (head == null) {
      print("Head is null.");
      return;
    }
    var temp = head;
    SinglyLinkedListNode? prev;
    while (temp != null && temp.data != data) {
      prev = temp;
      temp = temp.next;
    }
    if (temp != null) {
      if (prev == null) {
        head = temp.next;
      } else {
        prev.next = temp.next;
      }
      print("Found $data and removed. $head");
    } else {
      print("$data not found in $head");
    }
  }

  void get(int pos) {
    if (head == null) {
      print("Head is null.");
      return;
    }
    var tempPos = pos;
    var temp = head;
    while (temp != null && tempPos > 0) {
      --tempPos;
      temp = temp.next;
    }
    if (temp != null) {
      print("Found ${temp.data} at position $pos");
    } else {
      print("No element at position $pos");
    }
  }

  void set(int pos, int newData) {
    if (head == null) {
      print("Head is null.");
      return;
    }
    var tempPos = pos;
    var temp = head;
    while (temp != null && tempPos > 0) {
      --tempPos;
      temp = temp.next;
    }
    if (temp != null) {
      final oldData = temp.data;
      temp = SinglyLinkedListNode(data: newData);
      print("Found $oldData at position $pos and updated to $newData. $head");
    } else {
      print("No element at position $pos");
    }
  }

  void insert(int pos, int data) {
    final newNode = SinglyLinkedListNode(data: data);
    if (head == null) {
      head = newNode;
    } else {
      var temp = head;
      SinglyLinkedListNode? prev;
      var tempPos = pos;
      while ((temp != null || prev != null) && tempPos > 0) {
        --tempPos;
        prev = temp;
        temp = temp?.next;
      }
      if (prev == null && temp == null) {
        print("Position $pos is out of bounds");
        return;
      } else {
        if (prev == null) {
          head = newNode;
        } else {
          prev.next = newNode;
        }
        newNode.next = temp;
      }
    }
    print(
        "New node ${SinglyLinkedListNode(data: data)} inserted at position $pos in $head");
  }

  void removeAt(int pos) {
    if (head == null) {
      print("Head is null.");
      return;
    }
    var temp = head;
    SinglyLinkedListNode? prev;
    var tempPos = pos;
    while ((temp != null || prev != null) && tempPos > 0) {
      --tempPos;
      prev = temp;
      temp = temp?.next;
    }
    if (prev == null && temp == null) {
      print("Position $pos is out of bounds");
      return;
    } else {
      if (prev == null && temp != null) {
        head = temp.next;
      } else if (prev != null && temp == null) {
        pop();
      } else {
        prev?.next = temp?.next?.next;
      }
    }
    print("Item removed from position $pos in $head");
  }

  void reverse() {
    if (head == null) {
      print("Head is null.");
      return;
    }
    var first = head;
    SinglyLinkedListNode? second;
    SinglyLinkedListNode? third;
    while (first != null) {
      third = second;
      second = first;
      first = first.next;
      second.next = third;
    }
    if (second != null) {
      head = second;
    }
    print("Linked list reversed $head");
  }
}

File: kotlin/udemy_course/src/main/kotlin/data_structures/linked_lists/SinglyLinkedList.kt
package data_structures.linked_lists

class SinglyLinkedList {
    var head: ListNode? = null
        private set
    var tail: ListNode? = null
        private set
    var length = 0
        private set

    fun push(data: Int) {
        val listNode = ListNode(data = data)
        if (head == null && tail == null) {
            head = listNode
            tail = listNode
        } else {
            tail?.next = listNode
            tail = listNode
        }
        ++length
    }

    fun pop(): Int? {
        if (tail == null) {
            throw IllegalStateException("Linked List is Empty")
        }
        val data = tail?.data
        if (head == tail) {
            head = null
            tail = null
        } else {
            var temp = head
            while (temp?.next != tail) {
                temp = temp?.next
            }
            tail = temp
            tail?.next = null
        }
        --length
        return data
    }

    fun shift(): Int? {
        if (head == null) {
            throw IllegalStateException("Linked List is Empty")
        }
        val data = head?.data
        head = head?.next
        --length
        return data
    }

    fun unshift(data: Int) {
        val listNode = ListNode(data = data)
        listNode.next = head
        head = listNode
        if (tail == null) {
            tail = head
        }
        ++length
    }

    fun get(pos: Int): ListNode? {
        if (head == null) {
            throw IllegalStateException("Linked List is Empty")
        }
        if (pos < 0 || pos >= length) {
            throw IllegalArgumentException("Position should be >= 0 and <= $length")
        }

        var tempPos = pos
        var temp = head
        while (tempPos > 0 && temp != null) {
            temp = temp.next
            --tempPos
        }

        return temp
    }

    fun set(pos: Int, data: Int) {
        if (head == null) {
            throw IllegalStateException("Linked List is Empty")
        }
        if (pos < 0 || pos >= length) {
            throw IllegalArgumentException("Position should be >= 0 and <= $length")
        }

        var tempPos = pos
        var temp = head
        while (tempPos > 0 && temp != null) {
            temp = temp.next
            --tempPos
        }

        temp?.data = data
    }

    fun insert(pos: Int, data: Int) {
        if (head == null) {
            throw IllegalStateException("Linked List is Empty")
        }
        if (pos < 0 || pos >= length) {
            throw IllegalArgumentException("Position should be >= 0 and <= $length")
        }

        when (pos) {
            0 -> {
                unshift(data = data)
            }

            length - 1 -> {
                push(data = data)
            }

            else -> {
                val listNode = ListNode(data = data)

                var tempPos = pos
                var temp = head
                var prev: ListNode? = null

                while (tempPos > 0 && temp != null) {
                    prev = temp
                    temp = temp.next
                    --tempPos
                }

                listNode.next = temp
                prev?.next = listNode
                ++length
            }
        }
    }

    fun remove(pos: Int): Int? {
        if (head == null) {
            throw IllegalStateException("Linked List is Empty")
        }
        if (pos < 0 || pos >= length) {
            throw IllegalArgumentException("Position should be >= 0 and <= $length")
        }

        when (pos) {
            0 -> {
                return shift()
            }

            length - 1 -> {
                return pop()
            }

            else -> {
                var tempPos = pos
                var temp = head
                var prev: ListNode? = null

                while (tempPos > 0 && temp != null) {
                    prev = temp
                    temp = temp.next
                    --tempPos
                }

                prev?.next = temp?.next
                temp?.next = null

                --length

                return temp?.data
            }
        }
    }

    fun reverse() {
        if (head == null) {
            throw IllegalStateException("Linked List is Empty")
        }

        var temp = head
        head = tail
        tail = temp

        var next = temp?.next
        var prev: ListNode? = null

        while (temp != null) {
            temp.next = prev
            prev = temp
            temp = next
            next = next?.next
        }
    }

    fun print() {
        println("${head.toString()} | Length=$length")
    }
}